cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactRepository.java

```diff
@@ -0,0 +1,245 @@
+import com.google.inject.Inject;
+import io.cdap.cdap.api.artifact.ArtifactInfo;
+import io.cdap.cdap.api.artifact.ArtifactRange;
+import io.cdap.cdap.api.artifact.ArtifactSummary;
+import io.cdap.cdap.api.artifact.CloseableClassLoader;
+import io.cdap.cdap.api.plugin.PluginClass;
+import io.cdap.cdap.api.plugin.PluginSelector;
+import io.cdap.cdap.app.runtime.ProgramRunnerFactory;
+import io.cdap.cdap.common.ArtifactNotFoundException;
+import io.cdap.cdap.common.conf.ArtifactConfigReader;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.id.Id;
+import io.cdap.cdap.data2.metadata.writer.MetadataServiceClient;
+import io.cdap.cdap.internal.app.runtime.plugin.PluginNotExistsException;
+import io.cdap.cdap.internal.app.spark.SparkCompatReader;
+import io.cdap.cdap.proto.artifact.ApplicationClassInfo;
+import io.cdap.cdap.proto.artifact.ApplicationClassSummary;
+import io.cdap.cdap.proto.artifact.ArtifactSortOrder;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.security.impersonation.EntityImpersonator;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import javax.annotation.Nullable;
+
+/**
+ * RemoteArtifactRepository provides a remote implementation of ArtifactRepository
+ */
+public class RemoteArtifactRepository implements ArtifactRepository {
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteArtifactRepository.class);
+  private final ArtifactRepositoryReader artifactRepositoryReader;
+  private final ArtifactClassLoaderFactory artifactClassLoaderFactory;
+  private final ArtifactInspector artifactInspector;
+  private final Set<File> systemArtifactDirs;
+  private final ArtifactConfigReader configReader;
+  private final MetadataServiceClient metadataServiceClient;
+  private final Impersonator impersonator;
+
+  @VisibleForTesting
+  @Inject
+  public RemoteArtifactRepository(CConfiguration cConf, ArtifactRepositoryReader artifactRepositoryReader,
+                                  MetadataServiceClient metadataServiceClient,
+                                  ProgramRunnerFactory programRunnerFactory,
+                                  Impersonator impersonator) {
+    this.artifactRepositoryReader = artifactRepositoryReader;
+    this.artifactClassLoaderFactory = new ArtifactClassLoaderFactory(cConf, programRunnerFactory);
+    this.artifactInspector = new ArtifactInspector(cConf, artifactClassLoaderFactory);
+    this.systemArtifactDirs = new HashSet<>();
+    String systemArtifactsDir = cConf.get(Constants.AppFabric.SYSTEM_ARTIFACTS_DIR);
```

Comment:
Duplicated logic as in DefaultArtifactRepository. Please refactor appropriately.

Reply:
Removed unused fields in the class, this significantly cleaned up the constructor